src/cudamatrix/cu-kernels.cu

```diff
@@ -2629,19 +2629,109 @@ static void _diff_log_softmax(const MatrixDim in_deriv_dim,
   }
 }
 
-/// Add spatial regularization term to `out_deriv`.
+/// This CUDA kernel implements the idea 'spatial smoothing' as described in
+/// the paper https://arxiv.org/pdf/1610.05256v1.pdf
+///
+/// It adds spatial regularization term to `out_deriv`.
 /// The regularization term is calculated from `out_value`.
-/// Each row of `out_value` is first viewed as a 16*n grid,
-/// convoluted with the 3x3 kernel, scaled with `scale`
+/// Each row of `out_value` is first viewed as a n*16 or n*32 2D image,
+/// high-pass filtered by a 3x3 kernel, scaled with `scale`
```

Comment:
also, what do you mean by n*16 or n*32?  Your code seems to refer to 16.  What does this depend on?  And how about end effects?  I don't see from your code where the 32 comes in.
What is the dimension of regularization_sqsum?
Mention that out_deriv has the same dimension as out_value, but possibly a different stride.